File: modules/mastodon/items/media_attachment.py
from ..models import MediaAttachment
from .base import Item

MEDIA_TYPES = {
    0: "image",
    1: "gifv",
    2: "audio",
    3: "unknown",
    4: "video",
}


class MediaAttachmentItem(Item):
    id: str
    type: str
    url: str | None
    preview_url: str | None
    remote_url: str | None
    meta: dict | None
    description: str | None
    blurhash: str | None

    @classmethod
    def from_model(cls, media_attachment: MediaAttachment):
        return cls(
            id=str(media_attachment.id),
            type=MEDIA_TYPES[media_attachment.type],
            url=media_attachment.file_url,
            preview_url=media_attachment.preview_url,
            remote_url=media_attachment.remote_url,
            meta=media_attachment.file_meta,
            description=media_attachment.description,
            blurhash=media_attachment.blurhash,
        )

File: crates/mastodon/src/media_attachment.rs
use serde::Serialize;

/// Mastodon-compatible `MediaAttachment` entity.
/// See: <https://docs.joinmastodon.org/entities/MediaAttachment/>
#[derive(Debug, Serialize)]
pub struct MediaAttachment {
    pub id: String,
    #[serde(rename = "type")]
    pub media_type: String,
    pub url: Option<String>,
    pub preview_url: Option<String>,
    pub remote_url: Option<String>,
    pub meta: Option<MediaMeta>,
    pub description: Option<String>,
    pub blurhash: Option<String>,
}

/// Mastodon-compatible media metadata.
#[derive(Debug, Serialize)]
pub struct MediaMeta {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub original: Option<MediaMetaDimensions>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub small: Option<MediaMetaDimensions>,
}

/// Width/height/aspect for a media variant.
#[derive(Debug, Serialize)]
pub struct MediaMetaDimensions {
    pub width: u32,
    pub height: u32,
    pub aspect: f64,
}

impl MediaMetaDimensions {
    pub fn new(width: u32, height: u32) -> Self {
        let aspect = if height > 0 {
            f64::from(width) / f64::from(height)
        } else {
            1.0
        };
        Self {
            width,
            height,
            aspect,
        }
    }
}

pub fn normalize_media_type(t: &str) -> String {
    match t {
        "image" | "video" | "gifv" | "audio" => t.to_string(),
        _ => "unknown".to_string(),
    }
}
